### PR TITLE
[1.4.x] Bug fix in nullable constraints

### PR DESCRIPTION
**Pull request by aneeshafedo**
Resolves https://github.com/ballerina-platform/openapi-tools/issues/1290

## Approach

Avoid generating constraints when the type is nullable

Ex:

```yaml
Scores:
      type: array
      items:
        type: number
        nullable: true
      maxItems: 2
      minItems: 2
      nullable: true
      description: At least one entry should be non-null.
```

**Generated bal type**

```bal
# At least one entry should be non-null.
public type Scores decimal[]?;
```

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java [hunk 1]
```diff
@@ -262,24 +265,41 @@ private static RecordFieldWithDefaultValueNode getRecordFieldWithDefaultValueNod
...
+
+        boolean isConstraintSupport = schema.getNullable() != null && schema.getNullable() ||
+                (schema instanceof ComposedSchema && (((ComposedSchema) schema).getOneOf() != null ||
+                        ((ComposedSchema) schema).getAnyOf() != null));
+        boolean nullable = GeneratorMetaData.getInstance().isNullable();
+        if (nullable) {
+            return false;
+        } else if (isConstraintSupport) {
+            OUT_STREAM.printf("WARNING: constraints in the OpenAPI contract will be ignored for the " +
+                    "typ `%s`%n", typeName);
```
Comment (lnash94) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java:298]:
```suggestion
                    "type `%s`%n", typeName);
```

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java [hunk 2]
```diff
@@ -90,6 +90,9 @@ public class TypeGeneratorUtils {
             new ArrayList<>(Arrays.asList(GeneratorConstants.INTEGER, GeneratorConstants.NUMBER,
                     GeneratorConstants.STRING, GeneratorConstants.BOOLEAN));
 
+    public static final PrintStream OUT_STREAM = System.err;
+
```
Comment (lnash94) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java:94]: Additional new line

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java [hunk 3]
```diff
@@ -262,24 +265,41 @@ private static RecordFieldWithDefaultValueNode getRecordFieldWithDefaultValueNod
...
+            // Ignore Object, Map and Composed schemas.
+            return null;
         }
-        // Ignore Object, Map and Composed schemas.
         return null;
     }
 
+    public static boolean isSupportConstraint(String typeName, Schema schema) {
+
+        boolean isConstraintSupport = schema.getNullable() != null && schema.getNullable() ||
```
Comment (lnash94) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java:290]: Shall we change the `isConstraintSupport` variable to something like this ex: `constraintNotFollowed`, `avoidsConstraint` to align with the condition checks?

Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java:290]: Normally in Java it is a best practice to start a boolean variable with `is`. At the same time I was reusing the name defined [here](https://github.com/ballerina-platform/openapi-tools/blob/364cd5d8f34321af31c7b5283e7793530552d35c/openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java#L175) for the same purpose.

Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java]: Changed to `isConstraintAllowed`

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java [hunk 4]
```diff
@@ -134,4 +134,19 @@ public void testAdditionalPropertiesWithConstraint() throws IOException, Balleri
     }
     //TODO current tool doesn't handle union type: therefore union type constraint will handle once union type
     // generation available in tool.
+
+    @Test
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java:138]: Lets add the test description

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java [hunk 5]
```diff
@@ -262,24 +264,42 @@ private static RecordFieldWithDefaultValueNode getRecordFieldWithDefaultValueNod
...
+                return generateArrayConstraint(arraySchema);
+            }
+            // Ignore Object, Map and Composed schemas.
+            return null;
         }
-        // Ignore Object, Map and Composed schemas.
         return null;
     }
 
+    public static boolean allowsConstraints(String typeName, Schema schema) {
```
Comment (NipunaRanasinghe) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java:287]:
```suggestion
    public static boolean isConstraintAllowed(String typeName, Schema schema) {
```

File: openapi-cli/src/test/resources/generators/schema/ballerina/constraint/constraint_with_nullable.bal [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+public type Average float?;
+
+public type Pet record {
+    int id;
+    string name;
+    decimal[]? tag?;
+};
+
+public type Name string?;
```
[the remark this answers is not in the document]
Comment (aneeshafedo) [reply openapi-cli/src/test/resources/generators/schema/ballerina/constraint/constraint_with_nullable.bal:12]: There are few `oneOf` and `anyOf` usecases covered in the test `testRecordFiledConstraintWithZeroValue` in the same file

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java (hunk 4 above)
Comment (aneeshafedo) [reply openapi-cli/src/test/java/io/ballerina/openapi/generators/schema/ConstraintTests.java]: Added

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java (hunk 5 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java]: Updated

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java (hunk 3 above)
Comment (aneeshafedo) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/schema/TypeGeneratorUtils.java]: Changed to `isConstraintNotAllowed` in https://github.com/ballerina-platform/openapi-tools/pull/1301/commits/5b1a80e8798f333b69f746f2c5bb6162f9792b73

**Review by lnash94 — APPROVED** [review]

**Pull request merged**